### PR TITLE
WL#20856 Added Parenting Hotkeys to Edit.js Among Visual tweaks

### PR DESCRIPTION
**Pull request by Menithal**
https://worklist.net/20856

- Moved Audioscope hotkey from Modifiers + P to Modifiers + F2
- Added Window.notify to notifications.js for Generic non-error notifications (NotificationType.UNKNOWN)
- Added New Edit.js Advanced Menu options for Parenting / Unparenting Entities
- Added Hotkeys for Ctrl + P Parent and Ctrl + Shift + P to Unparent Entities. This should work even if the Edit Window is selected
- Added notification when User parents and unparents Entities via hotkeys
- Added Visual Indicators to Last Selected Entity - Last Entity will be bright yellow instead of orange
- Automatic Trailing Whitespace clean up on save of entitySelectionTool.js (sorry, this was accidental, I can revert this if needed)

To Test:

1. Rez a few entities, and open up edit mode.
2. Select the multiple entities. The Last entity you select should have a Yellow box instead of orange
3. Use  Ctrl - P to Parent objects
4. Observe Top Right popup
5. Check that entities are parented
6. Use Ctrl - Shift- P to Unparent entities
7. Observe Top Right popup
8. Check that entities are unparented
9. Repeat 3-8 while you select the edit window in an position
9. Repeat 3-8 with Advanced Menu only

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
An admin must verify this patch for the pull request build to fire. Type "ok to test" to begin testing this pull request or "add to whitelist" to test this and all future pull requests from this collaborator.

**Comment by ZappoMan** [comment]
add to whitelist

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7a5665ff4f2370f1cae6b018cb2568cd07a3aa59 available at: https://gist.github.com/ccf63aef464800b95b7847d06f6fb73a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-7a5665f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-7a5665f.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: afbf4bc30f53cd63efc2df1ab5ebfa14de7ec65d available at: https://gist.github.com/ccf63aef464800b95b7847d06f6fb73a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-afbf4bc.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-afbf4bc.dmg

**Comment by Nex-Pro** [comment]
Testing this.

**Comment by Nex-Pro** [comment]
Tested in Win 10 desktop mode.

Audio scope is now activated/de-activated using alt+F2.

parenting test:

Rez a few entities, and open up edit mode. **DONE**

Select the multiple entities. The Last entity you select should have a Yellow box instead of orange **SUCCESS**

Use Ctrl - P to Parent objects **DONE**

Observe Top Right popup  **SUCCESS** 

Check that entities are parented **SUCCESS**

Use Ctrl - Shift- P to Unparent entities **SUCCESS** 

Observe Top Right popup  **SUCCESS**

Check that entities are unparented **SUCCESS**
Repeat 3-8 with Advanced Menu **SUCCESS**

Note: There is an issue with moving the entities window which can cause this to fail.

To repro:
    Select several entities.
    reposition the entities pane.
    move mouse out of enties pane.
    press ctrl + p
    Observe that parenting fails.
    Also happens with unparent.
    Side effect is that the view swapped between first and third person.
    See attached images.

Would be good to have some way of doing this in HMD. Say some way of simulating holding shift.

## My view nicely zoomed out.

hifi-snap-by--on-2017-02-16_12-58-10
## After moving entity pane slightly and pressing CTRL+P
hifi-snap-by--on-2017-02-16_12-58-17
## After pressing CTRL+P again. I've lost my nice zoomed out view.
hifi-snap-by--on-2017-02-16_12-58-22

**Comment by Menithal** [comment]
Thanks,

Ill do a quick run through after I get home to nail down the hotkeys not being detected when focused on the entity Panel. I am guessing the QML instance there is actually grabbing the input.

**Comment by hifi-gustavo** [comment]
Build console output for commit: f211f015d7579ac476f1e1b863745730580c6be7 available at: https://gist.github.com/ccf63aef464800b95b7847d06f6fb73a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-f211f01.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-f211f01.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f211f015d7579ac476f1e1b863745730580c6be7 available at: https://gist.github.com/ccf63aef464800b95b7847d06f6fb73a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-f211f01.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-f211f01.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f5d266a56253703e35a8baf78159479ccac22ac7 available at: https://gist.github.com/ccf63aef464800b95b7847d06f6fb73a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-f5d266a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-f5d266a.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: dcbe3c622db92d2e5729813392a9c483c5ae1c22 available at: https://gist.github.com/ccf63aef464800b95b7847d06f6fb73a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-dcbe3c6.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-dcbe3c6.dmg

**Comment by Menithal** [comment]
I updated the scripts and you should now be able to parent entities even if the entity window is open and active.

**Comment by Nex-Pro** [comment]
### FAIL

Using build ending in 3c6
Can only do parenting operations with advanced menus turned on.

Problem with unintentionally switching between first and third person persists.

Good point:  Contrast between selected child ents and the parent is now much more obvious. 
Note. This does not actually show the parent it merely highlights the last selected entity.

**Comment by Menithal** [comment]
@Nex-Pro Darn it. I may need to just remove the hotkeys from the menu instead. The main issue I tried to correct in the last one was removed the main js call as I was experienced double parenting, will revert that change.

**Comment by hifi-gustavo** [comment]
Build console output for commit: f3c806e91090c720e4f3841600f7933d44258a7e available at: https://gist.github.com/ccf63aef464800b95b7847d06f6fb73a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-f3c806e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-f3c806e.dmg

**Comment by Menithal** [comment]
Should be Fixed now.

**Comment by Nex-Pro** [comment]
@Menithal  Ok Mr Fox. Checking now.

**Comment by Nex-Pro** [comment]
Waiting for re-build following menu order change.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2a4b493de560049bf34f2e179cca4970a830a7c7 available at: https://gist.github.com/ccf63aef464800b95b7847d06f6fb73a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-2a4b493.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-2a4b493.dmg

**Comment by Nex-Pro** [comment]
Got it.

**Comment by Nex-Pro** [comment]
QA Test: #9684
Testing platform: Win10 Desktop 16gb ram gtx1070 GPU.
Desktop mode.
HMD: Not required.

Moved Audioscope hotkey from Modifiers + P to Modifiers + F2: **SUCCESS**

Added Window.notify to notifications.js for Generic non-error notifications (NotificationType.UNKNOWN)

Added New Edit.js Advanced Menu options for Parenting / Unparenting Entities

Added Hotkeys for Ctrl + P Parent and Ctrl + Shift + P to Unparent Entities. This should work even if the Edit Window is selected

Added notification when User parents and unparents Entities via hotkeys

Added Visual Indicators to Last Selected Entity - Last Entity will be bright yellow instead of orange

Automatic Trailing Whitespace clean up on save of entitySelectionTool.js (sorry, this was accidental, I can revert this if needed)

To Test:

Rez a few entities, and open up edit mode. **DONE**

Select the multiple entities. The Last entity you select should have a Yellow box instead of orange: **SUCCESS**

Use Ctrl - P to Parent objects: **DONE**

Observe Top Right popup: **Popup displays message: ** ***"Entities Parented***

Check that entities are parented: **SUCCESS**

Use Ctrl - Shift- P to Unparent entities: **DONE**

Observe Top Right popup: **Partial FAIL. Popup displays message: ** ***"Entities Unarented*** **even when they were not.**

Check that entities are unparented: **SUCCESS**

Repeat 3-8 while you select the edit window in an position: **Results as above**

Repeat 3-8 with Advanced Menu only: **Results as above**


### Notes:

Displaying message informing user that "Entities Unparented" when they were not previously parented might lead to some frustration.
View mode swapping between first and third person persists as discussed.

**Comment by hifi-gustavo** [comment]
Build console output for commit: d9f86612de72cb10004d70ae68db4e06669650fb available at: https://gist.github.com/ccf63aef464800b95b7847d06f6fb73a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-d9f8661.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-d9f8661.dmg

**Comment by Nex-Pro** [comment]
### PASS

Tried to break it, couldn't. Fulfils all the above requirements. Nice work. And much needed.

**Comment by hifi-gustavo** [comment]
Build console output for commit: b1639ee3aafb07c96e1e516c74446bbe99737a9e available at: https://gist.github.com/ccf63aef464800b95b7847d06f6fb73a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-b1639ee.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-b1639ee.dmg

File: scripts/system/html/js/entityProperties.js [hunk 1]
```diff
@@ -1441,7 +1441,15 @@ function loaded() {
             }));
         });
 
-
+        document.addEventListener("keydown", function (keyDown) {
+          if (keyDown.keyCode === 80 && keyDown.ctrlKey) {
```
Comment (sethalves) [inline scripts/system/html/js/entityProperties.js]: Could we put all these 80s into a constant?

File: scripts/system/html/js/gridControls.js [hunk 2]
```diff
@@ -131,10 +131,17 @@ function loaded() {
 
         EventBridge.emitWebEvent(JSON.stringify({ type: 'init' }));
     });
-
+    document.addEventListener("keydown", function (keyDown) {
+      if (keyDown.keyCode === 80 && keyDown.ctrlKey) {
```
Comment (sethalves) [inline scripts/system/html/js/gridControls.js]: 80

File: scripts/system/edit.js [hunk 3]
```diff
@@ -1324,8 +1406,16 @@ Controller.keyReleaseEvent.connect(function (event) {
             });
             grid.setPosition(newPosition);
         }
+    } else if (event.key === 80 && event.isControl && !event.isAutoRepeat ) {
```
Comment (sethalves) [inline scripts/system/edit.js]: 80

File: scripts/system/html/js/entityList.js [hunk 4]
```diff
@@ -292,8 +292,15 @@ function loaded() {
               EventBridge.emitWebEvent(JSON.stringify({ type: 'delete' }));
               refreshEntities();
           }
+          if (keyDownEvent.keyCode === 80 && keyDownEvent.ctrlKey) {
```
Comment (sethalves) [inline scripts/system/html/js/entityList.js]: 80

**Comment by Menithal** [comment]
Will do!

File: scripts/system/html/js/entityProperties.js (hunk 1 above)
Comment (Menithal) [inline scripts/system/html/js/entityProperties.js]: Sure will do! 

Ideally however, all of the constants should be in a single configuration, but without module loading we cant really do it :<. For now I'll define them in each file as a constant KEY_P (as its the keycode for the character P)

**Comment by hifi-gustavo** [comment]
Build console output for commit: 91fa871d55cbc24f8f33d6dea102e9c1fb679977 available at: https://gist.github.com/ccf63aef464800b95b7847d06f6fb73a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-91fa871.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9684-91fa871.dmg

**Comment by Menithal** [comment]
@sethalves changes complete.

**Pull request merged**